### DISCUSSION TITLE
Thoughts on PocketBase/Drift

### DISCUSSION THREAD

**Discussion by rodydavis** [cachapa/sqlite_crdt#2] [General]
I currently build and use the package [pocketbase_drift](https://github.com/rodydavis/pocketbase_drift) which is a dart native client for [pocketbase](https://pocketbase.io/) and uses [drift](https://pub.dev/packages/drift) as the Sqlite3 layer.

I ended up doing something very similar to a CRDT for the sync (deleted, timestamp, synced) which was possible with the added columns.

As I am looking into this package more I think that this might be a good fit. But would like to make a drift_crdt first or sqlite3_crdt client first. [Sqlite3](https://pub.dev/packages/sqlite3) being an FFI based package.

Curious about your thoughts on this in regards to pocketbase being the source of truth on the server running SQLite, having hooks and can be extended as a framework. Looking through the code I want to find a way to call getChangeSet for a user and only sync specific tables and ones only needed for the user. It wasnt clear to me if the [crdt_sync](https://pub.dev/packages/crdt_sync) was syncing the whole database to every peer or just the user data.

The applications I am working on are all offline first and have highly connected peers.

**Comment by cachapa** [comment]
I'm not very familiar with PocketBase, but it seems to be an end-to-end  solution for data storage and synchronization in the style of Firebase?

The issue I see is that CRDTs require all nodes to follow the same ruleset for updating and merging changesets - including the server.

> Curious about your thoughts on this in regards to pocketbase being the source of truth on the server

The advantage of a CRDT is that there is no need for a "source of truth" since every node merges changesets deterministically, i.e. the result of a merge operation is always the same regardless of node or order of operations.
What you're suggesting is the traditional approach of having a server be the sole responsible for merging and conflict resolution.
This has the drawback of using plain timestamps and first-to-post-wins which is less than ideal in distributed applications.

What this all means is that it doesn't make much sense to have a client use HLC timestamps if the server cannot understand them.

> Looking through the code I want to find a way to call getChangeSet for a user and only sync specific tables and ones only needed for the user. It wasnt clear to me if the [crdt_sync](https://pub.dev/packages/crdt_sync) was syncing the whole database to every peer or just the user data.

By default both crdt_sync and [sql_crdt](https://github.com/cachapa/sql_crdt) (the parent package for sqlite_crdt) will return all tables in the dataset.
You can change this though:
- sql_crdt allows specifying [custom queries](https://github.com/cachapa/sql_crdt/blob/master/lib/src/sql_crdt.dart#L71) so you can e.g. filter by user id.
- crdt_sync has a changesetBuilder callback that triggers every time a changeset is sent to a client.

The tudo app has a [complex example of this](https://github.com/cachapa/tudo_server/blob/master/lib/tudo_server.dart#L17-L47), since it needs to send all items that are shared with that user.

It should be noted that a subset of a CRDT is itself a valid CRDT, so there should be no concerns merging different nodes even if some are missing part of the "whole picture".

> As I am looking into this package more I think that this might be a good fit. But would like to make a drift_crdt first or sqlite3_crdt client first. [Sqlite3](https://pub.dev/packages/sqlite3) being an FFI based package.

This package uses[ sqflite_common_ffi](https://pub.dev/packages/sqflite_common_ffi) which depends on sqlite3, so creating a sqlite3_crdt probably isn't worth the effort.

All in all I'm quite curious if you find any of this useful. I believe I have built a very robust sync framework for distributed apps, but it's currently missing a lot of tooling to make it generally useful - things like a login system, backend management UI, and other things that PocketBase and other solutions do very well.

Perhaps extending PocketBase to use [crdt] instead of their current DB layer would be one solution, though it's unclear to me how much effort it would be.

Finally, I would advise caution with using Sqlite in the backend if you're expecting you service to grow, as it doesn't scale very well.
I tried it when I released [Libra 4](https://libra-app.eu) with the cloud service and my server quickly buckled under pressure.
That is part of the reason why I started working on [postgres_sql](https://github.com/cachapa/postgres_crdt), which has been ticking fine even as the service grew over the past year.

**Reply by rodydavis** [reply]
Thank you so much for the feedback!

I am going to look into extending pocketbase and see how far I can get. PocketBase has a nice backend console and great client libraries too. There are some projects taking PocketBase and sharding it horizontally.

https://community.fly.io/t/fly-io-marmot-pocketbase-distributed-pocketbase/14129

I think there is a good connection between using CRDTs on the client and PocketBase for trusted nodes or backup.

I also found this package after learning about [GunDB](https://gun.eco/) and then finding a Dart client.

https://pub.dev/packages/flutter_gundb

**Reply by dragonballa** [reply]
@cachapa can you please tell me more about your libra 4 setup with sqlite? what was the issue? im surprised sqlite is problematic as ive been considering using Turso

also does this handle schema changes? is it propagated?

what about writing? is it talking to only one sqlite on the server and then replicating it out to other sqlite nodes (on other servers) ?

is this project still being worked on?

**Reply by dragonballa** [reply]
@rodydavis what were your experiences with gundb curious to know

**Reply by rodydavis** [reply]
Never got GunDB to production but was interesting in development.

**Reply by cachapa** [reply]
> @cachapa can you please tell me more about your libra 4 setup with sqlite? what was the issue? im surprised sqlite is problematic as ive been considering using Turso

My issues were specifically on the backend. Libra uses Sqlite on the mobile side and that works perfectly.

On the backend the issue was that Sqlite is single-threaded which was causing bottlenecks when multiple clients were connecting simultaneously.

TBF, I have made a number of performance improvements in the project since then so it might work better now, but I haven't tested it again.

The other issue with Sqlite on the backend is that you don't have the tooling available for e.g. Postgres. For example, I can simply connect to the production server with an SQL client to perform maintenance operations or gather user statistics.

> also does this handle schema changes? is it propagated?

Schema changes need to be manually applied on all nodes otherwise changesets won't merge cleanly.

I haven't gotten to the point where I would be able to do those automatically, and honestly I'm not sure it's possible without severely limiting the flexibility.

> what about writing? is it talking to only one sqlite on the server and then replicating it out to other sqlite nodes (on other servers) ?

In my particular case it's only one server running a Postgres DB behind the CRDT wrapper.

CRDTs don't really have the concept of "clients" and "servers", it's all nodes replicating data bidirectionally. In practice you don't want to replicate the entire backend to clients so clients effectively only receive the subset of the CRDT that is relevant to them.

You could have multiple "servers" which would simply be nodes syncing the full CRDT between each other.

> is this project still being worked on?

Yes, the lack of activity is simply because it is stable and feature complete, at least according to my requirements.
